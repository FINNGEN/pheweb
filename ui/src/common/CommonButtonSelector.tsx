import React, {ReactElement, useState} from "react";

interface SelectorProps<T> {
    options: T[];
    onSelect: (value: T | null) => void;
    initial?: (options: T[]) => T; // Function to compute initial value
    emptyElement?: ReactElement;
}

const defaultEmptyElement = <p className="text-gray-500">No options available</p>

const CommonButtonSelector = <T extends string | number>({
                                                             options,
                                                             onSelect,
                                                             initial = (opts) => opts[0], // Default: first element
                                                             emptyElement = defaultEmptyElement
                                                         }: SelectorProps<T>) => {
    const computedInitial = options.length > 0 ? initial(options) : null;
    const [selected, setSelected] = useState<T | null>(computedInitial);

    const handleClick = (value: T) => {
        setSelected(value);
        onSelect(value);
    };

    if (options.length === 0) {
        return emptyElement;
    }

    return (
        <div className="flex gap-2">
            {options.map((option) => (
                <button
                    key={option}
                    className={`px-4 py-2 rounded border ${
                        selected === option ? "bg-gray-200" : "bg-blue-500 text-white"
                    }`}
                    onClick={() => handleClick(option)}
                >
                    {option}
                </button>
            ))}
        </div>
    );
};

export default CommonButtonSelector;