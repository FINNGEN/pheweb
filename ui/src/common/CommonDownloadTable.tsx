import React, { ReactElement, useEffect, useState } from "react";
import ReactTable from "react-table-v6";
import commonLoading from "./CommonLoading";
import { createCSVLinkHeaders } from "./commonTableColumn";
import { CSVLink } from "react-csv";
import { Column, SortingRule } from "react-table";


/* Abstraction is broken here see:
 * https://stackoverflow.com/questions/53504924/reactjs-download-csv-file-on-button-click
 */
interface Link extends HTMLAnchorElement { link : HTMLAnchorElement }
interface Table<RowType> extends  JSX.Element { getResolvedState : () => { sortedData : RowType[] } }

export interface Props <TableData, RowType extends  {},
                        ReactProperties extends {} = {},
                        LinkProperties extends {} = {}>{
  filename : string
  tableData : (TableData| null)
  dataToTableRows : (data : TableData| null) => RowType[]
  tableColumns : Column<RowType>[]
  tableProperties? : ReactProperties,
  linkProperties? : LinkProperties,
  defaultSorted : SortingRule<string>[],
  subComponent? : ( JSX.Element | any ),
  tableRowToDownloadRow? : (columns : RowType[]) => RowType[] /* default to identity */
}

export type DownloadTableProps<TableData,
                               RowType extends  {},
                               ReactProperties extends {} = {}> = Props<TableData, RowType, ReactProperties>

const CommonDownloadTable = <TableData,RowType extends {}>
  ({ filename,
     tableData,
     dataToTableRows ,
     tableColumns ,
     tableProperties,
     linkProperties,
     defaultSorted,
     subComponent,
     tableRowToDownloadRow = (columns) => columns
   } : Props<TableData, RowType>) => {

    const [download, setDownload] = useState<RowType[] | null>(null);
    const [reactTableRef, setReactTableRef] = useState<Table<RowType> | null>(null);
    const [link, setLink] = useState<Link | null>(null);

   /* Using an effect because the data in tsv wasn't populated
    * when the link was clicked.
    */
    useEffect(() => { download && link && link.link.click() },[download,link]);

    const downloadHandler = () => {
      if (reactTableRef != null) {
        setDownload(tableRowToDownloadRow(reactTableRef.getResolvedState().sortedData))
      }
    }
    const body = <div>
      <ReactTable
        ref={setReactTableRef}
        data={dataToTableRows(tableData)}
        filterable
        columns={tableColumns}
        defaultSorted={defaultSorted}
        SubComponent={subComponent}
        {...tableProperties  } />

      <p>

      </p>
      <div className="row">
        <div className="col-xs-12">
          <div className="btn btn-primary" onClick={downloadHandler}>Download table</div>
        </div>
      </div>

      <CSVLink
        headers={createCSVLinkHeaders(tableColumns)}
        data={download == null ? [] : download }
        separator={"\t"}
        enclosingCharacter={""}
        filename={ filename }
        className="hidden"
        ref={(r) => setLink(r)}
        target="_blank"
        {...(linkProperties && {linkProperties : linkProperties})}
      />
    </div>
    const component = tableData == null? commonLoading : body;
    return component;
}

export default CommonDownloadTable
