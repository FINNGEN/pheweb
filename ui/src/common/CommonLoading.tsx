import React, { useState } from "react";

const commonLoading: JSX.Element = <div>... loading ...</div>

export const isLoading = (isLoading: boolean, content: () => JSX.Element): JSX.Element => {
  if (isLoading) {
    return commonLoading;
  } else {
    return content();
  }
}

export const CommonLoadingMoreOrLess = (qc_variant_results) => {
  const [showMore, setShowMore] = useState(false);

  const toggleColumns = () => {
    setShowMore(prev => !prev);
  };
  return <>
    <div style={{ color: '#007bff', textDecoration: 'none', cursor: 'pointer' }} onClick={toggleColumns}>
      {showMore ? 'Show Less' : 'Show More'}
    </div>
    <div style={showMore ? { display: 'block' } : { display: 'none' }}>
      {showMore ? (
        <pre style={{ backgroundColor: '#f6f8fa', padding: '1rem', borderRadius: '5px' }}>
          <code>{JSON.stringify(qc_variant_results, null, 2)}</code>
        </pre>
      ) : ''}
    </div>
  </>
}
const isEmpty = (obj: object) => Object.keys(obj).length === 0;

export const CommonQCVariantErrorTable = (qcVariantMessage: string | null | undefined, content: JSX.Element, qcVariantResults: object | null): JSX.Element => {
  const tableStyle = {
    width: '100%',
    borderCollapse: 'collapse' as 'collapse',  // Correctly specifying the value type
    margin: '16px 0'
  };


  return <div>
    {(qcVariantMessage === null || qcVariantMessage === undefined) ? (
      content
    ) : (
      <div>
        <b>Error!</b>: {qcVariantMessage}
        {isEmpty(qcVariantResults) ? <></> :
          (
            <>
              <table style={tableStyle}>
                <thead>
                  <tr>
                    <th>Variant</th>
                    <th>Failed filter</th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td>{qcVariantResults['variant']}</td>
                    <td>{qcVariantResults['failed_filter']}</td>
                  </tr>
                </tbody>
              </table>
              <CommonLoadingMoreOrLess qc_variant={qcVariantResults} />
            </>
          )
        }
      </div>
    )}
  </div>
}

export const hasError = (errorMessage: string | null | undefined, content: JSX.Element): JSX.Element => {
  if (errorMessage === null || errorMessage === undefined) {
    return content
  } else {
    return <div><b>Error</b> : {errorMessage}</div>
  }
}
export default commonLoading
