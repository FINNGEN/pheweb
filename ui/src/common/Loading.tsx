import React from "react";

const loading:  JSX.Element = <div>... loading ...</div>

export const isLoading = (isLoading : boolean, content: () =>  JSX.Element) :  JSX.Element => {
  if (isLoading) {
    return loading;
  } else {
    return content();
  }
}

export const hasError = (errorMessage : string | null, content:  JSX.Element) :  JSX.Element => {
  if(errorMessage == null){
    return content
  } else {
    return <div><b>Error</b> : {errorMessage}</div>
  }
}
export default loading
