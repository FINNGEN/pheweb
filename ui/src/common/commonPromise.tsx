type logger = (v: any) => void
export const promiseFulfilled = <X,>(r :  PromiseSettledResult<X>) : r is PromiseFulfilledResult<X> => r.status == "fulfilled";
export const logRejected : <X,>(r :  PromiseSettledResult<X>, logger?: logger) => PromiseSettledResult<X> = (r,logger =console.error) => { if(r.status == 'rejected') { logger(r.reason) }; return r; }
export const promiseFulfilledValue =  <X,>(v : PromiseFulfilledResult<X>) => v?.value;
export const promiseValues = <X,>(values :  PromiseSettledResult<X>[],logger?: logger) => values.map(x =>logRejected(x, logger)).filter(promiseFulfilled).map(promiseFulfilledValue);
export const notBottom = <X,>(x : X | null): x is X => x !== null && x !== undefined;
