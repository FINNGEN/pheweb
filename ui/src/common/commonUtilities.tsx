import React from "react";
import * as Handlebars from "handlebars/dist/cjs/handlebars";

/**
 * Compose fuction
 *
 * see : https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html
 *
 * @param f
 * @param g
 */
export function compose<A, B, C>(
  f: (arg: A) => B,
  g: (arg: B) => C
): (arg: A) => C {
  return (x) => g(f(x));
}

export type Handler = (url : string) => (e : Error) => void
const defaultHandler : Handler = (url : string) => (e : Error) => {
  warn(url,e)
}

/**
 * Get url
 *
 * Takes a setter and an optional transformation and a fetchURL
 * Fetches url as a json object.  Calls the sink with the resulting
 * json.  If there is an error it's sent to the console.
 *
 * @param url
 * @param sink
 * @param fetchURL
 */
export const get: <X>(url: string,
                      sink: (x: X) => void,
                      handler? :  Handler) => Promise<void> = (
  url,
  sink,
  handler = defaultHandler
) =>
  fetch(url)
    .then((response) => {
      if (response.status === 200) {
        return response.json();
      } else {
        const msg = `${response.statusText}`;
        throw new Error(msg);
      }
    })
    .then(sink)
    .catch(handler(url));

/**
 * Delete url
 *
 * Takes a setter and an optional transformation and a fetchURL
 * Fetches url as a json object.  Calls the sink with the resulting
 * json.  If there is an error it's sent to the console.
 *
 * @param url
 * @param sink
 * @param fetchURL
 */


export const deleteRequest : <X>(url: string,
                                 sink: (x: X) => void,
                                 handler? :  Handler) => Promise<void> = (
    url,
    sink,
    handler = defaultHandler
) =>
    fetch(url, { method : 'DELETE' })
        .then((response) => {
          if (response.status === 200) {
            return response.json();
          } else {
            throw new Error(response.statusText);
          }
        })
        .then(sink)
        .catch(handler(url));


/**
 * mustacheDiv
 *
 * return a div that contains as it's inner html
 * the output of a mustache template expanded using
 * the given context
 */
export const mustacheDiv: <X>(template: string, content: X) => JSX.Element = (
  template,
  content
) => (
  <div
    dangerouslySetInnerHTML={{ __html: Handlebars.compile(template)(content) }}
  ></div>
);

/**
 * mustacheSpan
 *
 * return a span that contains as it's inner html
 * the output of a mustache template expanded using
 * the given context
 */
export const mustacheSpan: <X>(template: string, content: X) => JSX.Element = (
  template,
  content
) => (
  <span
    dangerouslySetInnerHTML={{ __html: Handlebars.compile(template)(content) }}
  ></span>
);

/**
 * mustacheText
 *
 * return a span that contains as it's inner html
 * the output of a mustache template expanded using
 * the given context
 */
export const mustacheText: <X>(template: string, content: X) => string = (
  template,
  content
) => Handlebars.compile(template)(content);


export const fatal = (msg : string, context : {} ={}) : never =>{
  console.log(msg);
  alert(msg);
  console.log(context);
  throw msg;
}

export const warn = (msg : string, context : {} ={}) : void =>{
  console.warn(msg);
  console.log(context);
}
/**
 * flatten a list.
 * note this assumes expressions are well-typed and excludes null and undefined.
 * switch to .flat when upgrade to js 2019.
 */
export const flatten : <X>(a : X[][]) => X[] = (l) => l.reduce((a, v) =>a.concat(v),[]);
export const defaultEmptyArray : <X,Y>(a : X[],defaultArray :Y[]) => (X|Y)[] = (a,defaultArray) => a.length == 0?defaultArray:a;
