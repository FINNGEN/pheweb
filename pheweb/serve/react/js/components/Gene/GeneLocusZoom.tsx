import { mustacheDiv } from "../../common/Utilities";
import React from "react";
import { ConfigurationWindow } from "../Configuration/configurationModel";

const default_banner: string = `
<h3>Gene Summary</h3>
`
declare let window: ConfigurationWindow;
const { config } = window;
const banner: string = config?.userInterface?.gene?.locusZoom?.banner || default_banner;

interface Props {}
const GeneLocusZoom = (props : Props) => {

  return <div>
    {mustacheDiv(banner, [])}
  </div>
}

export default GeneLocusZoom;