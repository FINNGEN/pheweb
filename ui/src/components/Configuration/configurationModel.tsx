import { NotFoundConfiguration } from '../NotFound/notFoundModel'
import { ChipConfiguration } from '../Chip/chipModel'
import { CodingConfiguration } from '../Coding/codingModel'
import { IndexConfiguration } from '../Index/indexModel'
import { PhenotypeConfiguration } from '../Phenotype/phenotypeModel'
import { AboutConfiguration } from '../About/aboutModel'
import { VariantConfiguration } from '../Variant/variantModel'
import { TopHitsConfiguration } from '../TopHits/topHitsModel'
import { Gene } from '../Gene/geneModel'
import { LOFConfiguration } from '../LOF/lofModel'
import { Region } from '../Region/RegionModel'


export interface VisConfiguration {
    readonly info_tooltip_threshold : number,
    readonly loglog_threshold : number,
    readonly manhattan_colors : string[]
    readonly tooltip_template : string
}

export interface ApplicationConfiguration {
    readonly root? : string
    readonly title: string
    readonly logo : string
    readonly vis_conf : VisConfiguration
    readonly ld_panel_version : string
    readonly genome_build : number
    readonly browser: string
    readonly ld_service: string
    readonly ld_server : string

}

export interface ConfigurationUserInterface {
    application? : ApplicationConfiguration
    notFound? : NotFoundConfiguration
    chip? : ChipConfiguration
    index?: IndexConfiguration
    phenotype?: PhenotypeConfiguration
    about?: AboutConfiguration
    variant? : VariantConfiguration
    topHits? : TopHitsConfiguration
    gene? : Gene.Configuration
    lof? : LOFConfiguration
    coding? : CodingConfiguration
    region? : Region.Configuration
}
export interface ConfigurationMetaData {
    genome_build : number
}

export interface ConfigurationWindow extends  Window {
    config? : { userInterface? : ConfigurationUserInterface
                metaData? : ConfigurationMetaData
                application? : ApplicationConfiguration
    }
}

declare let window: ConfigurationWindow;

export const searchParams = (params : { [ key : string ] : (string | string[]) }) : string => {
       const urlSearchParams = new URLSearchParams()
        const keys = Object.keys(params).sort()
	keys.forEach(key => {
	   const value = params[key];
            if(Array.isArray(value)){
                value.forEach(v => urlSearchParams.append(key,v));
            } else {
                urlSearchParams.append(key,value);
            }
        })
	return urlSearchParams.toString();
}

export const resolveURL = (relativeURL : string,params : { [ key : string ] : (string | string[]) } | undefined = undefined) : string => {
    const root = window?.config?.application?.root
    const url = new URL(relativeURL,root?root:window.location.origin)
    if(params !== undefined){
        url.search = searchParams(params);
    }
    return url.toString()
}