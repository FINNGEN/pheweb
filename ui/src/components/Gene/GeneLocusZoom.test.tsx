/* eslint-env jest */
import {assoc_fields, association_layout, lz_conf} from './GeneLocusZoom';
import {GcLambda, Phenotype} from "../../common/commonModel";
import {Layout} from "locuszoom";
import {ConfigurationWindow} from "../Configuration/configurationModel";


test('check assoc fields binary', () => {
    const expected = [
        "association:maf_cases",
        "association:maf_controls",
        "association:id",
        "association:chr",
        "association:position",
        "association:ref",
        "association:alt",
        "association:pvalue",
        "association:pvalue|neglog10_or_100",
        "association:mlogp",
        "association:beta",
        "association:sebeta",
        "association:rsid",
        "association:maf",
        "association:most_severe",
        "association:fin_enrichment",
        "association:INFO",
        "ld:state",
        "ld:isrefvar",
        ];
    const phenotype : Phenotype = ({} as unknown as Phenotype);
    expect(assoc_fields(phenotype)).toStrictEqual(expected);
});

test('check assoc fields quant ', () => {
    const expected = [
        "association:id",
        "association:chr",
        "association:position",
        "association:ref",
        "association:alt",
        "association:pvalue",
        "association:pvalue|neglog10_or_100",
        "association:mlogp",
        "association:beta",
        "association:sebeta",
        "association:rsid",
        "association:maf",
        "association:most_severe",
        "association:fin_enrichment",
        "association:INFO",
        "ld:state",
        "ld:isrefvar",
        ];
    const phenotype : Phenotype = ({ is_binary : false }  as unknown as Phenotype);
    expect(assoc_fields(phenotype)).toStrictEqual(expected);
});

const samplePhenotype = {
    assoc_files: [] ,
    category: "None" ,
    category_index: 10 ,
    phenocode: "T2D" ,
    phenostring: "Type II"
};
test('check assoc layout', () => {
    expect(association_layout(samplePhenotype)).toMatchSnapshot()

});
