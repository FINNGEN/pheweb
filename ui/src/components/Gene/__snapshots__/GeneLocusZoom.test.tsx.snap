// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check assoc layout 1`] = `
Object {
  "axes": Object {
    "x": Object {
      "extent": "state",
      "label": "Chromosome {{chr}} (Mb)",
      "label_function": "chromosome",
      "label_offset": 32,
      "render": true,
      "tick_format": "region",
    },
    "y1": Object {
      "label": "-log10 p-value",
      "label_function": null,
      "label_offset": 28,
      "render": true,
    },
  },
  "background_click": "clear_selections",
  "dashboard": Object {
    "components": Array [
      Object {
        "color": "green",
        "position": "right",
        "type": "toggle_legend",
      },
    ],
  },
  "data_layers": Array [
    Object {
      "id": "significance",
      "offset": 7.301029995663981,
      "orientation": "horizontal",
      "type": "orthogonal_line",
    },
    Object {
      "behaviors": Object {
        "onclick": Array [
          Object {
            "action": "link",
            "href": "/variant/{{association:chr}}-{{association:position}}-{{association:ref}}-{{association:alt}}",
          },
        ],
        "onmouseout": Array [
          Object {
            "action": "unset",
            "status": "selected",
          },
        ],
        "onmouseover": Array [
          Object {
            "action": "set",
            "status": "selected",
          },
        ],
      },
      "color": Array [
        Object {
          "field": "ld:isrefvar",
          "parameters": Object {
            "field_value": 1,
            "then": "#9632b8",
          },
          "scale_function": "if",
        },
        Object {
          "field": "ld:state",
          "parameters": Object {
            "breaks": Array [
              0,
              0.2,
              0.4,
              0.6,
              0.8,
            ],
            "values": Array [
              "#357ebd",
              "#46b8da",
              "#5cb85c",
              "#eea236",
              "#d43f3a",
            ],
          },
          "scale_function": "numerical_bin",
        },
        "#B8B8B8",
      ],
      "fields": Array [
        "association:maf_cases",
        "association:maf_controls",
        "association:id",
        "association:chr",
        "association:position",
        "association:ref",
        "association:alt",
        "association:pvalue",
        "association:pvalue|neglog10_or_100",
        "association:mlogp",
        "association:beta",
        "association:sebeta",
        "association:rsid",
        "association:maf",
        "association:most_severe",
        "association:fin_enrichment",
        "association:INFO",
        "ld:state",
        "ld:isrefvar",
      ],
      "fill_opacity": 0.7,
      "id": "associationpvalues",
      "id_field": "association:id",
      "legend": Array [
        Object {
          "class": "lz-data_layer-scatter",
          "color": "#9632b8",
          "label": "LD Ref Var",
          "shape": "diamond",
          "size": 40,
        },
        Object {
          "class": "lz-data_layer-scatter",
          "color": "#d43f3a",
          "label": "1.0 > r² ≥ 0.8",
          "shape": "circle",
          "size": 40,
        },
        Object {
          "class": "lz-data_layer-scatter",
          "color": "#eea236",
          "label": "0.8 > r² ≥ 0.6",
          "shape": "circle",
          "size": 40,
        },
        Object {
          "class": "lz-data_layer-scatter",
          "color": "#5cb85c",
          "label": "0.6 > r² ≥ 0.4",
          "shape": "circle",
          "size": 40,
        },
        Object {
          "class": "lz-data_layer-scatter",
          "color": "#46b8da",
          "label": "0.4 > r² ≥ 0.2",
          "shape": "circle",
          "size": 40,
        },
        Object {
          "class": "lz-data_layer-scatter",
          "color": "#357ebd",
          "label": "0.2 > r² ≥ 0.0",
          "shape": "circle",
          "size": 40,
        },
        Object {
          "class": "lz-data_layer-scatter",
          "color": "#B8B8B8",
          "label": "no r² data",
          "shape": "circle",
          "size": 40,
        },
      ],
      "namespace": Object {
        "default": "",
        "ld": "ld",
      },
      "point_shape": Object {
        "field": "ld:isrefvar",
        "parameters": Object {
          "else": "circle",
          "field_value": 1,
          "then": "diamond",
        },
        "scale_function": "if",
      },
      "point_size": Object {
        "field": "ld:isrefvar",
        "parameters": Object {
          "else": 40,
          "field_value": 1,
          "then": 80,
        },
        "scale_function": "if",
      },
      "tooltip": Object {
        "closable": false,
        "hide": Object {
          "and": Array [
            "unhighlighted",
            "unselected",
          ],
        },
        "html": "
  <strong>{{association:id}}</strong><br/>
  <strong>{{association:rsid}}</strong><br/>
  <strong>{{association:most_severe}}</strong><br/>
  <table>
  <tbody
    <tr>
       <td>phenotype</td>
       <td><strong>Type II</strong></td>
    </tr>
    <tr>
      <td>p-value</td>
      <td><strong>{{association:pvalue|scinotation}}</strong></td>
    </tr>
    <tr>
      <td>beta</td>
      <td><strong>{{association:beta}}</strong> ({{association:sebeta}})</td>
    </tr>
    <tr>
      <td>-log10(p)</td>
      <td><strong>{{association:mlogp|scinotation}}</strong></td>
    </tr>
    <tr>
      <td>MAF</td>
      <td><strong>{{association:maf|percent}}</strong></td>
    </tr>
    <tr>
      <td>MAF controls</td>
      <td><strong>{{association:maf_controls|percent}}</strong></td>
    </tr>
    <tr>
      <td>MAF cases</td>
      <td><strong>{{association:maf_cases|percent}}</strong><br></td>
    <tr>
      <td>FIN enrichment</td>
      <td><strong>{{association:fin_enrichment}}</strong></td>
    </tr>
    <tr>
      <td>INFO</td>
      <td><strong>{{association:INFO}}</strong></td>
    </tr>
  </tbody>
  </table>
  ",
        "show": Object {
          "or": Array [
            "highlighted",
            "selected",
          ],
        },
      },
      "transition": false,
      "type": "scatter",
      "x_axis": Object {
        "axis": 1,
        "field": "association:position",
      },
      "y_axis": Object {
        "axis": 1,
        "field": "association:mlogp",
        "floor": 0,
        "min_extent": Array [
          0,
          10,
        ],
        "upper_buffer": 0.1,
      },
    },
  ],
  "description": null,
  "id": "association",
  "inner_border": "rgb(210, 210, 210)",
  "interaction": Object {
    "drag_background_to_pan": true,
    "drag_x_ticks_to_scale": true,
    "drag_y1_ticks_to_scale": true,
    "drag_y2_ticks_to_scale": true,
    "scroll_to_zoom": true,
    "x_linked": true,
    "y1_linked": false,
    "y2_linked": false,
  },
  "legend": Object {
    "height": 138,
    "hidden": true,
    "label_size": 12,
    "orientation": "vertical",
    "origin": Object {
      "x": 55,
      "y": 40,
    },
    "padding": 5,
    "width": 91.66200256347656,
  },
  "margin": Object {
    "bottom": 40,
    "left": 50,
    "right": 50,
    "top": 10,
  },
  "min_height": 100,
  "min_width": 400,
  "origin": Object {
    "x": 0,
    "y": 0,
  },
  "proportional_height": 0.3,
  "proportional_origin": Object {
    "x": 0,
    "y": 0,
  },
  "title": Object {
    "text": "Type II",
    "x": 55,
    "y": 30,
  },
  "y_index": 0,
}
`;
